Have the global Propagators default to a no-op instance. (#721)
open-telemetry/opentelemetry-specification: specification/context/api-propagators.md

Implementations MAY provide global `Propagator`s for
each supported `Propagator` type.

If offered, the global `Propagator`s MUST default to a no-op instance.

> Users MUST call [Set Global Propagator](#set-global-propagator) to

Reviewer: Who is "Users" here? Does this sentence guarantee application developers that unless they manually set a propagator, they will have no-op? I.e. does this sentence forbid the SDK from changing the default propagator by calling SetGlobalPropagator itself?

Author: That's something interesting to consider - I'd personally keep things simple and require users to set the global propagators, just like they add `SpanProcessor`s. But I understand the idea behind having the SDK install a few sane defaults (`TraceContext` plus `CorrelationContext`, I'm thinking). Is this what you have in mind?

Reviewer: Yes that is what I have in mind. I don't think this PR needs to handle what the SDK should do, but at least it should be clear if the SDK is even allowed to do this.

Reviewer: Please clarify the term "users" here as by my above comments. I think I agree with this PR, but it's too vague right now.

Implementations MAY provide global `Propagator`s for
each supported `Propagator` type.

> If offered, the global `Propagator`s MUST default to a no-op instance.

Reviewer 2: @carlosalberto in this PR are you advocating for code cleanliness or scenario? This may be the root of the misunderstanding. I think myself, @yurishkuro and @bogdandrutu all advocating for a long term goal of having context ambiently propagated everywhere. It may be that the extensions you mentioned below would be the mechanism of this enablement, but I believe this context should be everywhere.

For example, to explain better what happens in .NET. - there is no context propagation when nothing is enabled. But even smaller level of logging enablement (which is default) will enable context propagation. We clearly recognize some people want to turn it off, But default is ON.

Author: Sorry for the lack of context. A week ago, in the Specification SIG call (at least @bogdandrutu was there), we came to an agreement on having the API with no-op Propagators by default (.Net will keep its current behavior, given the specific needs of how it's used in the .Net framework and how things are instrumented by default).

> But even smaller level of logging enablement (which is default) will enable context propagation

Oh, actually this is what I tried to convey when I mentioned `OpenTelemetry extensions` (same in spirit, although not quite the same).

cc @bogdandrutu @yurishkuro

Author: Ping @SergeyKanzhelev

In order to override the default value, an call to
[Set Global Propagator](#set-global-propagator) MUST be done

> by either SDK implementations, OpenTelemetry extensions or final application code.

Reviewer 2: it is unclear what the OpenTelemetrt extension is. What did you have in mind?

Author: Any (future) helper OpenTelemetry extension. This makes me believe I should remove such mention for now, in order to avoid confusion ;)

> In order to override the default value, an call to
> [Set Global Propagator](#set-global-propagator) MUST be done
> by either SDK implementations, OpenTelemetry extensions or final application code.

Reviewer 3: ```suggestion
Another propagator can be set up by installing an SDK or by directly calling
[Set Global Propagator](#set-global-propagator) from application code.
```

The SDK doesn't necessarily have to call that method on the API, it can be implemented differently.
I'd also drop the confusing OTel extensions wording - if it's not worded as a "MUST either this or that", other ways are permitted implicitly.

Author: @reyang @SergeyKanzhelev Any 'official' documentation we could mention here for the .NET case?

Author: SGTM.

> Note: OpenTelemetry .Net is an exception to this rule as his global

Reviewer: ```suggestion
Note: OpenTelemetry .NET is an exception to this rule as his global
```

> Note: OpenTelemetry .NET is an exception to this rule as his global

Reviewer 3: ```suggestion
Note: OpenTelemetry .NET is an exception to this rule, as its global
```

> `Propagator`s default to a set of specific `Propagator`s, including `TraceContext`.

Reviewer 3: It would be nice to reference the .NET documentation giving more details on that.

Author: @reyang @SergeyKanzhelev Any 'official' documentation we could mention here for the .NET case?

Author: Ping @SergeyKanzhelev @reyang

Reviewer 4: It is here https://docs.microsoft.com/dotnet/api/system.diagnostics.activitycontext.
